Fix Env.without_cmis

lpw25 opened the pull request:
`Env.without_cmis` prevents cmi files from being loaded whilst printing types. However, this can corrupt the state of the environment. In particular, it can cause the `EnvLazy.t` representing a module's components to be empty incorrectly.

For example, given these two modules:

```ocaml
(* foo.mli *)
module type S = sig

  type t = T

end
```

```ocaml
(* bar.mli *)
module M : Foo.S

module N : sig

  module O : sig

    type t = T

  end

end
```

Then you can normally access `Bar.M.T` in the toplevel:

```ocaml
$ ocaml -short-paths
        OCaml version 4.06.0+dev1-2017-06-23

# let x = Bar.M.T;;
val x : Bar.M.t = Bar.M.T
```

but if you first print an appropriately long type from `Bar` then you can't:

```ocaml
$ ocaml -short-paths
        OCaml version 4.06.0+dev1-2017-06-23

# let y = Bar.N.O.T;;
val y : Bar.N.O.t = Bar.N.O.T
# let x = Bar.M.T;;
Characters 8-15:
  let x = Bar.M.T;;
          ^^^^^^^
Error: Unbound constructor Bar.M.T
```

In 4.04 and 4.05 this can even cause an assertion failure:

```ocaml
$ ocaml -short-paths
        OCaml version 4.04.1

# let x = Bar.N.O.T;;
open Bar.M;;
let y = T;;
val x : Bar.N.O.t = Bar.N.O.T
# # Fatal error: exception File "typing/env.ml", line 1099, characters 64-70: Assertion failed
```

This patch fixes the issue by logging any `EnvLazy.t`s which are set to empty within `without_cmis` and then restoring them when `without_cmis` is finished.

lpw25 committed:
Fix Env.without_cmis

alainfrisch commented on typing/env.ml:
```diff
@@ -72,38 +72,79 @@ let error err = raise (Error err)
+  type undo =
+    | Nil
+    | Cons : ('a, 'b) t * 'a * undo -> undo
+
```
You can avoid the GADT here, either by keeping a list of undo functions (`fun () -> x := Thunk e`) or by using a parametric type for undo (`('a, 'b) undo`).

alainfrisch commented on typing/env.ml:
```diff
@@ -72,38 +72,79 @@ let error err = raise (Error err)
   val get_arg : ('a,'b) t -> 'a option
 
+  val log : unit -> log
+  val force_logged : log -> ('a -> 'b option) -> ('a,'b option) t -> 'b option
```
Perhaps add a small comment to explain the semantics of log/force_logged/backtrack.

alainfrisch commented:
The approach is coherent with the current approach, but all this logic is quite fragile.  It's not obvious from the code that no other memoized stuff depends on the "global" switch (the other use of EnvLazy is fine since it should be "pure" enough; I did not try to check carefully if the memoization tables in functor_components could be affected but I don't think so).  Similarly, it's not obvious that a `Not_found` raised by `find_pers_struct` (because of without_cmis) is the only reason for `components_of_module_maker` returning None.

Anyway, this fixes a clear bug and the risk of regression of small (at worst, this would be a performance regression in the printer), so I think this can be merged, after adding a non-regression test and a changelog entry.

lpw25 commented:
I agree this whole thing is fragile, but I'm not too worried as I have a complete reimplementation of short-paths that removes it and I'm hoping to get that into the compiler before too long (it's already being used on merlin's development branch but I'd like to tidy it up and document it before submitting it to the compiler -- which will probably be after the 4.06 freeze).

lpw25 committed:
Add regression test

lpw25 committed:
Add changes entry

lpw25 commented:
Thanks for the prompt review.

Ok, I've added a regression test and a changes entry. I've added a comment on `force_logged`, but I've left the GADT as it is since I prefer it to the other options. I've also given `can_use_cmis` it's own type rather than using `EnvLazy.log option` to make things a little clearer.

Octachron commented on typing/env.ml:
```diff
@@ -678,25 +757,24 @@ let read_pers_struct check modname filename =
-      in
-      add_import name;
-      acknowledge_pers_struct check name ps
+  | exception Not_found ->
```
Would it not make sense to merge the two conditions here and have 
`| exception Not_found when !can_load_cmis = Can_load_cmis -> ` rather than `reraise Not_found`
on the `Cannot_load_cmis` branch?

lpw25 replied:
I dislike using `=` for this kind of thing

Octachron replied:
I understand, thanks for humouring my question.

alainfrisch reviewed (commented):
LGTM

alainfrisch commented:
Considering this fixes an actual bug, I think that a changelog entry should be added (it will also make CI happy).

alainfrisch committed:
Merge branch 'trunk' into fix-without-cmis

alainfrisch commented:
(Sorry, I somehow missed the Changelog entry.  I've fixed a conflict, and will merge when the CI feedback is green.)

alainfrisch committed:
Merge branch 'trunk' into fix-without-cmis

alainfrisch commented:
AppVeyor failure is unrelared, merging.

shindere commented:
@lpw25: I am just working on migrating the typing-short-paths/gpr1223.ml
test to ocamltest.

The migration is not straightforward because, at the moment, ocamltest
in general and the toplevel test in particular can not deal with
.mli-only modules.

The tool can probably be extended to deal with such situations, but I am not
sure yet how this should be done and before starting to think aobut it more,
I am wondering whether this test really has to be as it is, or whether it
could perhaps be simplified somehow?

lpw25 commented:
I think you can just add some `.ml` files for them if you want. It shouldn't affect the test itself, which is `gpr1223.ml`.

shindere commented:
Leo White (2017/12/19 09:16 -0800):
> I think you can just add some `.ml` files for them if you want. It
> should affect the test, which is `gpr1223.ml`.

Do you mean it should _not_ affect the test?

Shall these .mlfiles duplicate the mli ones?

trefis commented:
> Do you mean it should _not_ affect the test?
> Shall these .mlfiles duplicate the mli ones?

I'm pretty sure the answer to both questions is "yes".

lpw25 commented:
Yes, I meant "should not" (and had corrected it in the comment, but obviously that doesn't change the email notifications).

Any ml file that can compile with those interfaces is fine -- the actual test only depends on the mli files.

shindere commented:
Leo White (2017/12/19 13:15 -0800):
> Yes, I meant "should not" (and had corrected it in the comment, but
> obviously that doesn't change the email notifications).

Ow, sorry, I did not even realise it was possible to correct comments.

The limits of not using GitHub's web interface... Sorry for the trouble.

lpw25 commented:
No problem, I'll try to make a new comment correcting myself in future so that you get an email for it.

shindere commented:
Leo White (2017/12/19 22:50 -0800):
> No problem, I'll try to make a new comment correcting myself in future
> so that you get an email for it.

Thanks. It'd be nice if GitHub had a setting to also receive
notifications for modifications, actually.
